File: Unknownbank/app/src/main/java/com/kasatria/kasatriaunknownbank/CreditCardExtraDetailsScreen.kt
package com.kasatria.kasatriaunknownbank

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

import com.kasatria.kasatriaunknownbank.ui.theme.PrimaryBlue
import com.kasatria.kasatriaunknownbank.ui.theme.TextPrimary
import com.kasatria.kasatriaunknownbank.ui.theme.White


@Composable
fun CreditCardExtraDetailsScreen(
    onBack: () -> Unit,
    onNext: () -> Unit
) {

    var monthlyIncome by rememberSaveable {
        mutableStateOf("12,000")
    }

    var monthlyCommitments by rememberSaveable {
        mutableStateOf("5,860")
    }

    var retirementIncome by rememberSaveable {
        mutableStateOf("Rental Income")
    }

    var statementDelivery by rememberSaveable {
        mutableStateOf("Email")
    }

    var collectionState by rememberSaveable {
        mutableStateOf(
            "Wilayah Persekutuan Kuala Lumpur"
        )
    }

    var collectionDistrict by rememberSaveable {
        mutableStateOf("Kuala Lumpur")
    }

    var branch by rememberSaveable {
        mutableStateOf("Bangsar")
    }


    Box(
        modifier = Modifier.fillMaxSize()
    ) {

        Image(
            painter = painterResource(
                id = R.drawable.full_background_light
            ),
            contentDescription = null,
            modifier = Modifier.fillMaxSize(),
            contentScale = ContentScale.Crop
        )


        Column(
            modifier = Modifier
                .fillMaxSize()
                .statusBarsPadding()
        ) {

            /*
             * =================================
             * HEADER
             * =================================
             */

            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(56.dp)
            ) {

                IconButton(
                    onClick = onBack,

                    modifier = Modifier
                        .padding(start = 8.dp)
                        .align(
                            Alignment.CenterStart
                        )
                ) {

                    Icon(
                        imageVector =
                            Icons.AutoMirrored.Filled.ArrowBack,

                        contentDescription = "Back",

                        tint = TextPrimary
                    )
                }


                Text(
                    text =
                        "Applying for Bank World Mastercard",

                    modifier = Modifier
                        .align(Alignment.Center)
                        .padding(
                            start = 55.dp,
                            end = 30.dp
                        ),

                    color = Color(0xFF333333),

                    fontSize = 13.sp,

                    fontWeight =
                        FontWeight.Bold,

                    maxLines = 1
                )
            }


            /*
             * =================================
             * PROGRESS
             * =================================
             */

            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(
                        start = 20.dp,
                        end = 23.dp,
                        top = 20.dp
                    ),

                horizontalArrangement =
                    Arrangement.SpaceBetween,

                verticalAlignment =
                    Alignment.CenterVertically
            ) {

                Text(
                    text =
                        "Some Extra Details Will Help",

                    color =
                        Color(0xFF333333),

                    fontSize = 15.sp,

                    fontWeight =
                        FontWeight.Bold
                )


                Text(
                    text = "80%",

                    color =
                        Color(0xFF0EAF5F),

                    fontSize = 14.sp,

                    fontWeight =
                        FontWeight.Medium
                )
            }


            Spacer(
                modifier =
                    Modifier.height(11.dp)
            )


            Box(
                modifier = Modifier
                    .padding(horizontal = 20.dp)
                    .fillMaxWidth()
                    .height(6.dp)
                    .background(
                        color =
                            Color(0xFFE6E8EA),

                        shape =
                            RoundedCornerShape(
                                100.dp
                            )
                    )
            ) {

                Box(
                    modifier = Modifier
                        .fillMaxWidth(0.80f)
                        .height(6.dp)
                        .background(
                            color =
                                Color(0xFF0EAF5F),

                            shape =
                                RoundedCornerShape(
                                    100.dp
                                )
                        )
                )
            }


            Spacer(
                modifier =
                    Modifier.height(23.dp)
            )


            /*
             * =================================
             * FORM
             * =================================
             */

            Column(
                modifier = Modifier
                    .weight(1f)
                    .verticalScroll(
                        rememberScrollState()
                    )
                    .padding(
                        start = 20.dp,
                        end = 20.dp,
                        bottom = 110.dp
                    ),

                verticalArrangement =
                    Arrangement.spacedBy(18.dp)
            ) {

                ExtraDetailsTextField(
                    label =
                        "Monthly Net Income (RM)",

                    value =
                        monthlyIncome,

                    onValueChange = {
                        monthlyIncome = it
                    }
                )


                ExtraDetailsTextField(
                    label =
                        "Other Monthly Commitments (RM)",

                    value =
                        monthlyCommitments,

                    onValueChange = {
                        monthlyCommitments = it
                    }
                )


                ExtraDetailsSelectionField(
                    label =
                        "Source of Income After Retirement",

                    value =
                        retirementIncome,

                    options = listOf(
                        "Rental Income",
                        "Savings",
                        "Investment",
                        "Pension",
                        "Business Income",
                        "Others"
                    ),

                    onSelected = {
                        retirementIncome = it
                    }
                )


                ExtraDetailsSelectionField(
                    label =
                        "Send My Statement to",

                    value =
                        statementDelivery,

                    options = listOf(
                        "Email",
                        "Mail"
                    ),

                    onSelected = {
                        statementDelivery = it
                    }
                )


                ExtraDetailsSelectionField(
                    label =
                        "Card Collection State",

                    value =
                        collectionState,

                    options = listOf(
                        "Wilayah Persekutuan Kuala Lumpur",
                        "Selangor",
                        "Johor",
                        "Penang",
                        "Perak"
                    ),

                    onSelected = {
                        collectionState = it
                    }
                )


                ExtraDetailsSelectionField(
                    label =
                        "Card Collection District",

                    value =
                        collectionDistrict,

                    options = listOf(
                        "Kuala Lumpur",
                        "Cheras",
                        "Sentul",
                        "Bukit Bintang"
                    ),

                    onSelected = {
                        collectionDistrict = it
                    }
                )


                ExtraDetailsSelectionField(
                    label =
                        "Select Branch",

                    value = branch,

                    options = listOf(
                        "Bangsar",
                        "KLCC",
                        "Bukit Bintang",
                        "Mid Valley"
                    ),

                    onSelected = {
                        branch = it
                    }
                )
            }
        }


        /*
         * =================================
         * NEXT
         * =================================
         */

        Button(
            onClick = onNext,

            modifier = Modifier
                .align(
                    Alignment.BottomCenter
                )
                .navigationBarsPadding()
                .padding(
                    start = 16.dp,
                    end = 16.dp,
                    bottom = 16.dp
                )
                .fillMaxWidth()
                .height(46.dp),

            shape =
                RoundedCornerShape(100.dp),

            colors =
                ButtonDefaults.buttonColors(
                    containerColor =
                        PrimaryBlue,

                    contentColor =
                        White
                )
        ) {

            Text(
                text = "Next",

                fontSize = 15.sp,

                fontWeight =
                    FontWeight.Bold
            )
        }
    }
}


@Composable
private fun ExtraDetailsTextField(
    label: String,
    value: String,
    onValueChange: (String) -> Unit
) {

    Column {

        Text(
            text = label,
            color = Color(0xFF333333),
            fontSize = 14.sp
        )


        Spacer(
            modifier =
                Modifier.height(7.dp)
        )


        BasicFormField(
            value = value,

            onValueChange =
                onValueChange,

            backgroundColor =
                White,

            modifier =
                Modifier.fillMaxWidth()
        )
    }
}


@Composable
private fun ExtraDetailsSelectionField(
    label: String,
    value: String,
    options: List<String>,
    onSelected: (String) -> Unit
) {

    Column {

        Text(
            text = label,
            color = Color(0xFF333333),
            fontSize = 14.sp
        )


        Spacer(
            modifier =
                Modifier.height(7.dp)
        )


        SelectionBox(
            value = value,

            options = options,

            onSelected = onSelected,

            backgroundColor =
                White,

            modifier =
                Modifier.fillMaxWidth()
        )
    }
}

File: unknownbank_flutter/lib/screens/credit_card_extra_details_screen.dart
import 'package:flutter/material.dart';

import '../models/credit_card_application_data.dart';
import '../theme/app_colors.dart';

class CreditCardExtraDetailsScreen
    extends StatefulWidget {
  final String productName;
  final CreditCardApplicationData applicationData;

  final ValueChanged<CreditCardApplicationData>
      onApplicationDataChange;

  final VoidCallback onNext;

  const CreditCardExtraDetailsScreen({
    super.key,
    required this.productName,
    required this.applicationData,
    required this.onApplicationDataChange,
    required this.onNext,
  });

  @override
  State<CreditCardExtraDetailsScreen> createState() =>
      _CreditCardExtraDetailsScreenState();
}

class _CreditCardExtraDetailsScreenState
    extends State<CreditCardExtraDetailsScreen> {
  late CreditCardApplicationData data;

  @override
  void initState() {
    super.initState();
    data = widget.applicationData;
  }

  void _update(
    CreditCardApplicationData updatedData,
  ) {
    setState(() {
      data = updatedData;
    });

    widget.onApplicationDataChange(updatedData);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stack(
        children: [
          Positioned.fill(
            child: Image.asset(
              'assets/images/full_background_light.png',
              fit: BoxFit.cover,
            ),
          ),

          SafeArea(
            child: Column(
              children: [
                SizedBox(
                  height: 56,
                  child: Stack(
                    alignment: Alignment.center,
                    children: [
                      Align(
                        alignment: Alignment.centerLeft,
                        child: Padding(
                          padding:
                              const EdgeInsets.only(left: 8),
                          child: IconButton(
                            onPressed: () {
                              Navigator.of(context).pop();
                            },
                            icon: const Icon(
                              Icons.arrow_back,
                              color:
                                  AppColors.textPrimary,
                            ),
                          ),
                        ),
                      ),

                      Padding(
                        padding: const EdgeInsets.only(
                          left: 55,
                          right: 30,
                        ),
                        child: Text(
                          'Applying for ${widget.productName}',
                          maxLines: 1,
                          overflow:
                              TextOverflow.ellipsis,
                          style: const TextStyle(
                            fontSize: 13,
                            fontWeight:
                                FontWeight.bold,
                            color:
                                Color(0xFF333333),
                          ),
                        ),
                      ),
                    ],
                  ),
                ),

                const Padding(
                  padding: EdgeInsets.fromLTRB(
                    20,
                    20,
                    23,
                    0,
                  ),
                  child: Row(
                    mainAxisAlignment:
                        MainAxisAlignment
                            .spaceBetween,
                    children: [
                      Text(
                        'Some Extra Details Will Help',
                        style: TextStyle(
                          fontSize: 15,
                          fontWeight:
                              FontWeight.bold,
                          color:
                              Color(0xFF333333),
                        ),
                      ),

                      Text(
                        '80%',
                        style: TextStyle(
                          fontSize: 14,
                          fontWeight:
                              FontWeight.w500,
                          color:
                              Color(0xFF0EAF5F),
                        ),
                      ),
                    ],
                  ),
                ),

                const SizedBox(height: 11),

                Padding(
                  padding:
                      const EdgeInsets.symmetric(
                    horizontal: 20,
                  ),
                  child: ClipRRect(
                    borderRadius:
                        BorderRadius.circular(100),
                    child:
                        const LinearProgressIndicator(
                      value: 0.80,
                      minHeight: 6,
                      backgroundColor:
                          Color(0xFFE6E8EA),
                      valueColor:
                          AlwaysStoppedAnimation(
                        Color(0xFF0EAF5F),
                      ),
                    ),
                  ),
                ),

                const SizedBox(height: 23),

                Expanded(
                  child: SingleChildScrollView(
                    padding:
                        const EdgeInsets.fromLTRB(
                      20,
                      0,
                      20,
                      110,
                    ),
                    child: Column(
                      children: [
                        _TextField(
                          label:
                              'Monthly Net Income (RM)',
                          value:
                              data.monthlyNetIncome,
                          keyboardType:
                              TextInputType.number,
                          onChanged: (value) {
                            _update(
                              data.copyWith(
                                monthlyNetIncome:
                                    value,
                              ),
                            );
                          },
                        ),

                        const SizedBox(height: 18),

                        _TextField(
                          label:
                              'Other Monthly Commitments (RM)',
                          value: data
                              .monthlyCommitments,
                          keyboardType:
                              TextInputType.number,
                          onChanged: (value) {
                            _update(
                              data.copyWith(
                                monthlyCommitments:
                                    value,
                              ),
                            );
                          },
                        ),

                        const SizedBox(height: 18),

                        _DropdownField(
                          label:
                              'Source of Income After Retirement',
                          value: data
                              .retirementIncomeSource,
                          options: const [
                            'Rental Income',
                            'Savings',
                            'Investment',
                            'Pension',
                            'Business Income',
                            'Others',
                          ],
                          onChanged: (value) {
                            _update(
                              data.copyWith(
                                retirementIncomeSource:
                                    value,
                              ),
                            );
                          },
                        ),

                        const SizedBox(height: 18),

                        _DropdownField(
                          label:
                              'Send My Statement to',
                          value:
                              data.statementDelivery,
                          options: const [
                            'Email',
                            'Mail',
                          ],
                          onChanged: (value) {
                            _update(
                              data.copyWith(
                                statementDelivery:
                                    value,
                              ),
                            );
                          },
                        ),

                        const SizedBox(height: 18),

                        _DropdownField(
                          label:
                              'Card Collection State',
                          value:
                              data.collectionState,
                          options: const [
                            'Wilayah Persekutuan Kuala Lumpur',
                            'Selangor',
                            'Johor',
                            'Penang',
                            'Perak',
                          ],
                          onChanged: (value) {
                            _update(
                              data.copyWith(
                                collectionState:
                                    value,
                              ),
                            );
                          },
                        ),

                        const SizedBox(height: 18),

                        _DropdownField(
                          label:
                              'Card Collection District',
                          value:
                              data.collectionDistrict,
                          options: const [
                            'Kuala Lumpur',
                            'Cheras',
                            'Sentul',
                            'Bukit Bintang',
                          ],
                          onChanged: (value) {
                            _update(
                              data.copyWith(
                                collectionDistrict:
                                    value,
                              ),
                            );
                          },
                        ),

                        const SizedBox(height: 18),

                        _DropdownField(
                          label: 'Select Branch',
                          value:
                              data.collectionBranch,
                          options: const [
                            'Bangsar',
                            'KLCC',
                            'Bukit Bintang',
                            'Mid Valley',
                          ],
                          onChanged: (value) {
                            _update(
                              data.copyWith(
                                collectionBranch:
                                    value,
                              ),
                            );
                          },
                        ),
                      ],
                    ),
                  ),
                ),
              ],
            ),
          ),

          Positioned(
            left: 16,
            right: 16,
            bottom: 16,
            child: SafeArea(
              top: false,
              child: SizedBox(
                height: 46,
                child: FilledButton(
                  onPressed: widget.onNext,
                  style: FilledButton.styleFrom(
                    backgroundColor:
                        AppColors.primary,
                    shape:
                        RoundedRectangleBorder(
                      borderRadius:
                          BorderRadius.circular(
                        100,
                      ),
                    ),
                  ),
                  child: const Text(
                    'Next',
                    style: TextStyle(
                      fontSize: 15,
                      fontWeight:
                          FontWeight.bold,
                    ),
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class _TextField extends StatefulWidget {
  final String label;
  final String value;
  final ValueChanged<String> onChanged;
  final TextInputType? keyboardType;

  const _TextField({
    required this.label,
    required this.value,
    required this.onChanged,
    this.keyboardType,
  });

  @override
  State<_TextField> createState() =>
      _TextFieldState();
}

class _TextFieldState extends State<_TextField> {
  late final TextEditingController controller;

  @override
  void initState() {
    super.initState();

    controller =
        TextEditingController(text: widget.value);
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment:
          CrossAxisAlignment.start,
      children: [
        Text(
          widget.label,
          style: const TextStyle(
            fontSize: 14,
            color: Color(0xFF333333),
          ),
        ),

        const SizedBox(height: 7),

        SizedBox(
          height: 46,
          child: TextField(
            controller: controller,
            keyboardType: widget.keyboardType,
            onChanged: widget.onChanged,
            style: const TextStyle(
              fontSize: 15,
              color: Color(0xFF666666),
            ),
            decoration: InputDecoration(
              filled: true,
              fillColor:
                  const Color(0xFFFFFFFF),
              contentPadding:
                  const EdgeInsets.symmetric(
                horizontal: 19,
              ),
              border: OutlineInputBorder(
                borderRadius:
                    BorderRadius.circular(8),
                borderSide:
                    const BorderSide(
                  color: Color(0xFFDDDDDD),
                ),
              ),
              enabledBorder:
                  OutlineInputBorder(
                borderRadius:
                    BorderRadius.circular(8),
                borderSide:
                    const BorderSide(
                  color: Color(0xFFDDDDDD),
                ),
              ),
            ),
          ),
        ),
      ],
    );
  }
}

class _DropdownField extends StatelessWidget {
  final String label;
  final String value;
  final List<String> options;
  final ValueChanged<String> onChanged;

  const _DropdownField({
    required this.label,
    required this.value,
    required this.options,
    required this.onChanged,
  });

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment:
          CrossAxisAlignment.start,
      children: [
        Text(
          label,
          style: const TextStyle(
            fontSize: 14,
            color: Color(0xFF333333),
          ),
        ),

        const SizedBox(height: 7),

        Container(
          height: 46,
          padding: const EdgeInsets.only(
            left: 19,
            right: 10,
          ),
          decoration: BoxDecoration(
            color: Colors.white,
            borderRadius:
                BorderRadius.circular(8),
            border: Border.all(
              color:
                  const Color(0xFFDDDDDD),
            ),
          ),
          child: DropdownButtonHideUnderline(
            child: DropdownButton<String>(
              value: value,
              isExpanded: true,
              icon: const Icon(
                Icons.keyboard_arrow_down,
                color: Color(0xFF777777),
              ),
              style: const TextStyle(
                color: Color(0xFF666666),
                fontSize: 15,
              ),
              items: options
                  .map(
                    (option) =>
                        DropdownMenuItem<String>(
                      value: option,
                      child: Text(option),
                    ),
                  )
                  .toList(),
              onChanged: (value) {
                if (value != null) {
                  onChanged(value);
                }
              },
            ),
          ),
        ),
      ],
    );
  }
}